session/filetransfer.go

```diff
@@ -0,0 +1,496 @@
+	}
+}
+
+func (h *FileTransferHandler) InitFileUpload(msg *ws.ProtoMsg, w ResponseWriter) (err error) {
+	var (
+		fd          *os.File
+		defaultMode uint32 = 0644
+		defaultUID  uint32 = uint32(os.Getuid())
+		defaultGID  uint32 = uint32(os.Getgid())
+	)
+	params := FileInfo{
+		UID:  &defaultUID,
+		GID:  &defaultGID,
+		Mode: &defaultMode,
+	}
+	defer func() {
+		if err != nil {
+			h.Error(msg, w, err)
+		}
+	}()
+	err = msgpack.Unmarshal(msg.Body, &params)
+	if err != nil {
+		return errors.Wrap(err, "malformed request body")
+	} else if err = params.Validate(); err != nil {
+		return errors.Wrap(err, "invalid request parameters")
+	}
+
+	select {
+	case h.mutex <- struct{}{}:
+		defer func() {
+			if err != nil {
+				<-h.mutex
+			}
+		}()
+	default:
+		err = errors.New("another file transfer is already in progress")
+		return err
+	}
+	fd, err = ioutil.TempFile(path.Dir(*params.Path), path.Base(*params.Path))
+	if err != nil {
+		return errors.Wrap(err, "failed to create file")
+	}
+	err = fd.Chown(int(*params.UID), int(*params.GID))
+	if err != nil {
+		return errors.Wrap(err, "failed to set file permissions")
+	}
+	rsp := *msg
+	rsp.Header.MsgType = wsft.MessageTypeACK
+	rsp.Body = nil
+	err = w.WriteProtoMsg(&rsp)
+	if err != nil {
+		log.Error("failed to respond to client")
+	}
+	go h.FileUploadHandler(fd, msg, params, w)
+	return nil
+}
+
+func (h *FileTransferHandler) FileUploadHandler(
+	fd *os.File,
+	msg *ws.ProtoMsg,
+	params FileInfo,
+	w ResponseWriter,
+) (err error) {
+	defer func() {
+		<-h.mutex
+		filename := fd.Name()
+		errClose := fd.Close()
+		if errClose != nil {
+			log.Warnf("error closing file: %s", err.Error())
+		}
+		if err != nil {
+			log.Error(err.Error())
+			h.Error(msg, w, err)
+			errRm := os.Remove(filename)
+			if errRm != nil {
+				log.Errorf(
+					"error removing file after aborting upload: %s",
+					errRm.Error(),
+				)
+			}
+		}
+	}()
+	var (
+		offset int64
+		i      int
+
+		done bool
+		open bool
+	)
+	writeChunk := func(msg *ws.ProtoMsg) error {
+		if msg.Header.MsgType != wsft.MessageTypeChunk {
+			return errors.New("internal error")
+		}
+		chunkOffset, ok := msg.Header.Properties["offset"].(int64)
+		if !ok {
+			return errors.New("invalid file chunk message: missing offset property")
+		} else {
+			if chunkOffset != offset {
+				return errors.New("received unexpected chunk offset")
+			}
+		}
+		if len(msg.Body) > 0 {
+			n, err := fd.Write(msg.Body)
+			offset += int64(n)
+			if err != nil {
+				return errors.Wrap(err, "failed to write file chunk")
+			}
+		} else {
+			// EOF
+			return io.EOF
+		}
+		return nil
+	}
+	for !done {
+	Loop:
+		for i = 0; i < ACKSlidingWindowSend; i++ {
+			select {
+			case msg, open = <-h.msgChan:
+				if !open {
+					return nil
+				}
+				err := writeChunk(msg)
+				if err == io.EOF {
+					done = true
+				} else if err != nil {
+					return err
+				}
+			default:
+				break Loop
+			}
+		}
+		if i == 0 {
```

Comment:
I do not understand how and when we can enter this if?

Reply:
Sorry, this is not really explained in the code. The above loop will consume all inbound messages that are readily available without blocking on I/O. The `i==0` is only the case if no messages are pushed on the channel, and we need to block for messages.

Reply:
I refactored this bit to make it a bit clearer.